Loki: Run query when pressing Enter on line-filters

svennergr opened the pull request:
**What this PR does / why we need it**:
Pressing the Enter-Key on the line-filters in the Loki QueryBuilder is currently doing nothing. This PR adds the missing call to `onRunQuery` in the `LokiQueryEditorSelector.onChangeInternal` method. 

After doing that this would run queries also when the line-filters are blurred. To overcome this I added an `onBlur` method, which does not call `onRunQuery` if the QueryBuilder is used in Explore.

**Which issue(s) this PR fixes**:

Fixes #49408

**Special notes for your reviewer**:
Video of the fix:

https://user-images.githubusercontent.com/8092184/171174744-8ecd715f-6512-4863-a51b-6f91938df0d1.mov

svennergr committed:
added `onBlur` handler in LokiQueryBuilder

- `onBlur` is passed on to the `AutoSizeInput` components

svennergr committed:
changed `onBlur` and `onKeyDown` handling

- `onCommitChange` is only called if `onBlur` or `onKeyDown` are not set

torkelo reviewed (commented):
This is changing things in a way that I am not sure is good. We specifically don't auto run the query on changes in this new builder, instead highlight the run query button when there is a change and the query has not run. 

But when pressing enter key in the line filters (and in some other inputs) I think it should run. So only when pressing enter (or shift enter). For this I think we need a new option on the OperationParamDef  ("runOnEnter" or something close to that). Then we can check that in the onCommitChange callback

torkelo commented on public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.tsx:
```diff
@@ -49,6 +49,13 @@ export const LokiQueryEditorSelector = React.memo<LokiQueryEditorProps>((props)
+    onRunQuery();
+  };
+
+  const onBlur = () => {
+    if (app !== CoreApp.Explore) {
+      onRunQuery();
+    }
```
I would like to keep the behavior the same in Explore & elsewhere

svennergr replied:
I see. Just to be sure about the behaviour in the line-filters input components: 
- onBlur should highlight the `Run Query` button
- onKeyDown with `Enter` will run the query depending on a property from `QueryBuilderOperationParamDef`.

torkelo commented on public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.tsx:
```diff
@@ -49,6 +49,13 @@ export const LokiQueryEditorSelector = React.memo<LokiQueryEditorProps>((props)
   const onChangeInternal = (query: LokiQuery) => {
     setDataIsStale(true);
     onChange(query);
+    onRunQuery();
```
```suggestion
```

svennergr committed:
Revert "added `onBlur` handler in LokiQueryBuilder"

This reverts commit 47dbf9da15a5e968bb0ab8e4fb8e84b2c597a009.

svennergr committed:
added `runQueryOnEnter` flag to OperationParamDef

svennergr committed:
run query if `runQueryOnEnter` is configured

svennergr commented:
Hi @torkelo,

I now added a `runQueryOnEnter` flag to the `QueryBuilderOperationParamDef` and added that flag to all those line-filters:
- Line contains
- Line does not contain
- Line contains regex match
- Line does not match regex

Are there any more operation types which should run on enter-press?

In the [`onCommitChange` of `AutoSizeInput`](https://github.com/grafana/grafana/pull/49913/files#diff-0be9584796826c55c8c8cedabfcaaecb327cd90202db28e81b23b7996aa22170R41-R43) I needed to check for the event type, to no run queries on blur.

Just one last thing I would like to highlight: 
With that change the behaviour differs a bit between the "Operations" and the "Options", which might be confusing for some users:
![image](https://user-images.githubusercontent.com/8092184/171356000-6d2921dd-ce00-4991-80a2-7ce209936a6e.png)

svennergr commented on public/app/plugins/datasource/prometheus/querybuilder/shared/OperationParamEditor.tsx:
```diff
@@ -38,6 +38,9 @@ function SimpleInputParamEditor(props: QueryBuilderOperationParamEditorProps) {
       title={props.paramDef.description}
       onCommitChange={(evt) => {
         props.onChange(props.index, evt.currentTarget.value);
+        if (props.paramDef.runQueryOnEnter && evt.type !== 'blur') {
```
But maybe it's better to directly check for `evt.type === 'keydown'` instead of the negative check.

svennergr committed:
changed evt.type check to `keydown`
